### PR TITLE
ARROW-5136: [Flight] Call options

### PR DESCRIPTION
**Pull request by ghost**
Right now the only option is a timeout on an overall call, which will be useful to limit hangs when connecting to a Flight service that might have gone down. (gRPC is rather aggressive about retrying for a while before giving up.)

You could imagine having a health check action in DoAction, for instance, which fails if the call doesn't complete within a short time.

I would also like a stream timeout, where the timeout only fires if there is a long delay between messages on a DoGet or DoAction, but this is hard to implement as-is in gRPC. Implementing an async client interface might be a better way to achieve this.

### CODE REVIEW

File: cpp/src/arrow/flight/client.h [hunk 1]
```diff
@@ -41,6 +42,29 @@ namespace flight {
 
 class ClientAuthHandler;
 
+/// \brief A duration type for Flight call timeouts.
+typedef std::chrono::duration<double, std::chrono::seconds::period> TimeoutDuration;
+
+/// \brief Hints to the underlying RPC layer for Arrow Flight calls.
```
Comment (pitrou) [inline cpp/src/arrow/flight/client.h]: If this is simply a container for configuration values, perhaps we can make it a light-weight `struct` with public fields and no implementation class?

**Review by pitrou — COMMENTED** [review]
I won't comment about the Java parts, just one suggestion about the C++ side.

**Comment by pitrou** [comment]
(also, can you rebase? :-))

**Comment by wesm** [comment]
@jacques-n can you have a look at the Java changes?

File: cpp/src/arrow/flight/client.h (hunk 1 above)
Comment (lihalite) [reply cpp/src/arrow/flight/client.h]: That makes sense.

**Comment by lihalite** [comment]
On vacation, will update and rebase when I get in on Tuesday!

**Comment by lihalite** [comment]
Rebased, made FlightCallOptions a POD type in C++.

File: cpp/src/arrow/flight/client.h [hunk 2]
```diff
@@ -41,6 +42,26 @@ namespace flight {
...
+class ARROW_EXPORT FlightCallOptions {
+ public:
+  /// Create a default set of call options.
+  FlightCallOptions();
+
+  /// \brief Set a timeout on this call. If the call does not complete
+  /// within the specified period, the call will complete with a
+  /// non-OK status.
+  void set_timeout(const TimeoutDuration& timeout_sec);
```
Comment (pitrou) [inline cpp/src/arrow/flight/client.h]: Is this method still useful?

Comment (lihalite) [reply cpp/src/arrow/flight/client.h]: Not really, it'll be more helpful to have some static factory methods instead. I'll make that change.

**Comment by lihalite** [comment]
Rebased with latest master.

**Comment by pitrou** [comment]
@lihalite Do you need someone to vouch for the Java changes? @jacques-n perhaps?

**Comment by lihalite** [comment]
@pitrou I would appreciate someone to review the Java changes, yes, if you're good on the C++/Python ones.

**Comment by pitrou** [comment]
The C++/Python changes look ok to me.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java [hunk 3]
```diff
@@ -86,35 +86,83 @@ public FlightClient(BufferAllocator incomingAllocator, Location location) {
...
+  }
+
+  /**
+   * Get a list of available flights.
+   *
+   * @param options RPC-layer hints for the call. May be null for default options.
+   * @param criteria Criteria for selecting flights
+   * @return FlightInfo Iterable
+   */
+  public Iterable<FlightInfo> listFlights(CallOptions options, Criteria criteria) {
```
Comment (jacques-n) [inline java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java]: I've been thinking about this further. I'm not a huge fan of the double methods. What do you think about changing this to using vararg to manage both:

e.g. listFlights(Criteria c, CallOption...options)

Then we can simply introduce new CallOption items in the future. It means a user can call without any options or a selection of options as needed without introducing multiple apis.

Thoughts?

**Comment by lihalite** [comment]
Thanks everyone for the feedback! I'll rebase my outstanding PRs and submit the next part of our proposals soon.

**Comment by pitrou** [comment]
And Thank you @lihalite for your contributions!

**Pull request closed without merging**